monash-human-power/dashboard
# Read Live Data

## Description

This PR allows `dashboard` to get the relevant data 'live' from both, the wireless module sensor (to display the `reedDistance`) and from `boost` (to display available configs and get the distance offset to calculate the calibrated distance). 

In a minority change, it also adds a success toast whenever calibration is `reset`.

This PR is based off #105 and requires that to be merged first.

Another thing to note: From my discussion with Chris, we decided that after a user sets a calibrated value and it's sent to `boost`, the `dashboard` would wait for `boost` to send back the distance offset and only then will the displayed calibrated distance change (i.e. it will not immediately update when the user sets the value). The main advantage of this is that both `boost` and `dashboard` will be in sync and there will only be one source of truth, although this is at the expense of a slight delay to the `dashboard` user's experience.

## Screenshots
Reading distance value from WM 3.
<img width="1266" alt="Screen Shot 2021-03-27 at 5 39 52 pm" src="https://user-images.githubusercontent.com/63642262/112743213-15bc2f00-8fe1-11eb-8bda-58eafdfe1276.png">

Reading the distance offset from `boost`:
<img width="1264" alt="Screen Shot 2021-03-27 at 5 37 51 pm" src="https://user-images.githubusercontent.com/63642262/112743288-b7dc1700-8fe1-11eb-95f9-e05d0e6120c8.png">

Reading available configs from `boost`:
<img width="1243" alt="Screen Shot 2021-03-27 at 5 37 35 pm" src="https://user-images.githubusercontent.com/63642262/112743294-c75b6000-8fe1-11eb-8869-9f8c0914a0ed.png">



## Steps to Test

1) Run the fake WM sensor script in the `data-aquisition-repo`, you only need WM 3
2) Run the server in `dashboard` and the client
3) Verify that the data published the DAS script is being published as the `distance` on the `dashboard`
4) Publish the following and see if the calibrated distance is updated to reflect the `distanceOffset` value sent:

> `mosquitto_pub -h localhost -t 'boost/recommended_sp' -m '{"power": 10, "speed": 20, "zoneDistance": 100, "distanceOffset": 10, "distanceLeft": 200}'`

5) Publish the names of the available configs:

> `mosquitto_pub -h localhost -t 'boost/configs' -m '{"rider": [{"displayName": "Default rider", "fileName": "default.json"}], "track": [{"displayName": "Holden", "fileName": "holden.json"}, {"displayName": "Ford", "fileName": "ford.json"}, {"displayName": "Battle Mountain", "fileName": "bm.json"}], "bike": [{"displayName": "Wombat", "fileName": "v2.json"}, {"displayName": "Wombat", "fileName": "v2_old_camera.json"}, {"displayName": "Priscilla", "fileName": "v3.json"}, {"displayName": "Blacksmith", "fileName": "v1.5.json"}], "powerPlan": [{"displayName": "test", "fileName": "test.json"}, {"displayName": "default", "fileName": "default.json"}]}'`

and verify that they are displayed on the `dashboard`

6) Click `Reset` and check out the toast!
